File: lld/week-02-03-patterns/state/python/state.py
"""State pattern — an Order delegates behavior to its current OrderState,
which decides what transition, if any, is legal from that state.

See ../README.md for the design writeup.
"""
from __future__ import annotations

from typing import Protocol


class IllegalTransitionError(Exception):
    pass


class OrderState(Protocol):
    name: str

    def pay(self, order: "Order") -> None: ...
    def ship(self, order: "Order") -> None: ...
    def deliver(self, order: "Order") -> None: ...
    def cancel(self, order: "Order") -> None: ...


class Order:
    """The context: holds a reference to its current state and forwards
    every action to it."""

    def __init__(self, order_id: str):
        self.id = order_id
        self.state: OrderState = CreatedState()

    def pay(self) -> None:
        self.state.pay(self)

    def ship(self) -> None:
        self.state.ship(self)

    def deliver(self) -> None:
        self.state.deliver(self)

    def cancel(self) -> None:
        self.state.cancel(self)


class CreatedState:
    name = "Created"

    def pay(self, order: Order) -> None:
        order.state = PaidState()

    def ship(self, order: Order) -> None:
        raise IllegalTransitionError("cannot ship an order that has not been paid")

    def deliver(self, order: Order) -> None:
        raise IllegalTransitionError("cannot deliver an order that has not shipped")

    def cancel(self, order: Order) -> None:
        order.state = CancelledState()


class PaidState:
    name = "Paid"

    def pay(self, order: Order) -> None:
        raise IllegalTransitionError("order already paid")

    def ship(self, order: Order) -> None:
        order.state = ShippedState()

    def deliver(self, order: Order) -> None:
        raise IllegalTransitionError("cannot deliver an order that has not shipped")

    def cancel(self, order: Order) -> None:
        order.state = CancelledState()


class ShippedState:
    name = "Shipped"

    def pay(self, order: Order) -> None:
        raise IllegalTransitionError("order already paid")

    def ship(self, order: Order) -> None:
        raise IllegalTransitionError("order already shipped")

    def deliver(self, order: Order) -> None:
        order.state = DeliveredState()

    def cancel(self, order: Order) -> None:
        raise IllegalTransitionError("cannot cancel an order that has already shipped")


class DeliveredState:
    name = "Delivered"

    def pay(self, order: Order) -> None:
        raise IllegalTransitionError("order already delivered")

    def ship(self, order: Order) -> None:
        raise IllegalTransitionError("order already delivered")

    def deliver(self, order: Order) -> None:
        raise IllegalTransitionError("order already delivered")

    def cancel(self, order: Order) -> None:
        raise IllegalTransitionError("cannot cancel a delivered order")


class CancelledState:
    name = "Cancelled"

    def pay(self, order: Order) -> None:
        raise IllegalTransitionError("order is cancelled")

    def ship(self, order: Order) -> None:
        raise IllegalTransitionError("order is cancelled")

    def deliver(self, order: Order) -> None:
        raise IllegalTransitionError("order is cancelled")

    def cancel(self, order: Order) -> None:
        raise IllegalTransitionError("order is already cancelled")


def _demo() -> None:
    order = Order("O-1")
    print(f"Initial state: {order.state.name}")
    order.pay()
    order.ship()
    order.deliver()
    print(f"Final state: {order.state.name}")


if __name__ == "__main__":
    _demo()

File: lld/week-02-03-patterns/state/go/state.go
// Package state demonstrates the State pattern: an Order delegates its
// behavior (Pay, Ship, Deliver, Cancel) to its current OrderState object,
// which decides what transition, if any, is legal from that state.
package state

import "errors"

var ErrIllegalTransition = errors.New("illegal transition for current order state")

// OrderState is the interface every concrete state implements. Each method
// returns the next state on success, or ErrIllegalTransition if the
// transition isn't allowed from that state.
type OrderState interface {
	Name() string
	Pay(o *Order) error
	Ship(o *Order) error
	Deliver(o *Order) error
	Cancel(o *Order) error
}

// Order is the context: it holds a reference to its current state and
// forwards every action to it.
type Order struct {
	ID    string
	state OrderState
}

func NewOrder(id string) *Order {
	return &Order{ID: id, state: CreatedState{}}
}

func (o *Order) State() OrderState { return o.state }

func (o *Order) Pay() error     { return o.transition(o.state.Pay) }
func (o *Order) Ship() error    { return o.transition(o.state.Ship) }
func (o *Order) Deliver() error { return o.transition(o.state.Deliver) }
func (o *Order) Cancel() error  { return o.transition(o.state.Cancel) }

func (o *Order) transition(action func(*Order) error) error {
	return action(o)
}

func (o *Order) setState(s OrderState) { o.state = s }

// CreatedState: order placed, awaiting payment. Can pay or cancel.
type CreatedState struct{}

func (CreatedState) Name() string { return "Created" }
func (CreatedState) Pay(o *Order) error {
	o.setState(PaidState{})
	return nil
}
func (CreatedState) Ship(o *Order) error    { return ErrIllegalTransition }
func (CreatedState) Deliver(o *Order) error { return ErrIllegalTransition }
func (CreatedState) Cancel(o *Order) error {
	o.setState(CancelledState{})
	return nil
}

// PaidState: payment captured, awaiting shipment. Can ship or cancel (refund).
type PaidState struct{}

func (PaidState) Name() string       { return "Paid" }
func (PaidState) Pay(o *Order) error { return ErrIllegalTransition }
func (PaidState) Ship(o *Order) error {
	o.setState(ShippedState{})
	return nil
}
func (PaidState) Deliver(o *Order) error { return ErrIllegalTransition }
func (PaidState) Cancel(o *Order) error {
	o.setState(CancelledState{})
	return nil
}

// ShippedState: in transit. Can only be delivered; cannot cancel once shipped.
type ShippedState struct{}

func (ShippedState) Name() string        { return "Shipped" }
func (ShippedState) Pay(o *Order) error  { return ErrIllegalTransition }
func (ShippedState) Ship(o *Order) error { return ErrIllegalTransition }
func (ShippedState) Deliver(o *Order) error {
	o.setState(DeliveredState{})
	return nil
}
func (ShippedState) Cancel(o *Order) error { return ErrIllegalTransition }

// DeliveredState: terminal success state. No further transitions.
type DeliveredState struct{}

func (DeliveredState) Name() string           { return "Delivered" }
func (DeliveredState) Pay(o *Order) error     { return ErrIllegalTransition }
func (DeliveredState) Ship(o *Order) error    { return ErrIllegalTransition }
func (DeliveredState) Deliver(o *Order) error { return ErrIllegalTransition }
func (DeliveredState) Cancel(o *Order) error  { return ErrIllegalTransition }

// CancelledState: terminal state. No further transitions.
type CancelledState struct{}

func (CancelledState) Name() string           { return "Cancelled" }
func (CancelledState) Pay(o *Order) error     { return ErrIllegalTransition }
func (CancelledState) Ship(o *Order) error    { return ErrIllegalTransition }
func (CancelledState) Deliver(o *Order) error { return ErrIllegalTransition }
func (CancelledState) Cancel(o *Order) error  { return ErrIllegalTransition }
